File: 2017-12/2017-12-22/leetcode714.java
class Solution {
    public int maxProfit(int[] prices, int fee) {
        int len = prices.length;
        int[] dp = new int[len];
        int cash = 0, hold = -prices[0];
        for(int i = 1; i < len; i++)
        {
            cash = Math.max(cash, hold + prices[i] - fee);
            hold = Math.max(hold, cash - prices[i]);
        }
        return cash;
    }
}

File: 2018-08/2018-08-06/leetcode714.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices, int fee) {
        int len = prices.size();
        if (len == 0) return 0;
        
        int cash = 0;   // the maximum profit if have no stock
        int hold = -prices[0];  // the maximum profit if have one stock
        for (int i = 1; i < len; i++)
        {
            cash = max(cash, prices[i] + hold - fee);
            hold = max(hold, cash - prices[i]);
        }
        return cash;
    }
};